homeassistant/components/modbus/validators.py

```diff
@@ -371,7 +236,122 @@ def register_int_list_validator(value: Any) -> Any:
 
 def check_config(config: dict) -> dict:
     """Do final config check."""
-    config2 = duplicate_modbus_validator(config)
-    config3 = scan_interval_validator(config2)
-    config4 = duplicate_entity_validator(config3)
-    return config4
+    hosts: set[str] = set()
+    hub_names: set[str] = set()
+    hub_name_inx = 0
+    minimum_scan_interval = 0
+    ent_names: set[str] = set()
+    ent_addr: set[str] = set()
+
+    def validate_modbus(hub: dict) -> bool:
+        """Validate modbus entries."""
+        nonlocal hub_name_inx
+        host: str = (
+            hub[CONF_PORT]
+            if hub[CONF_TYPE] == SERIAL
+            else f"{hub[CONF_HOST]}_{hub[CONF_PORT]}"
+        )
+        if CONF_NAME not in hub:
+            hub[CONF_NAME] = (
+                DEFAULT_HUB if not hub_name_inx else f"{DEFAULT_HUB}_{hub_name_inx}"
+            )
+            hub_name_inx += 1
+            err = f"Modbus host/port {host} is missing name, added {hub[CONF_NAME]}!"
+            _LOGGER.warning(err)
+        name = hub[CONF_NAME]
+        if host in hosts or name in hub_names:
+            err = f"Modbus {name} host/port {host} is duplicate, not loaded!"
+            _LOGGER.warning(err)
+            return False
+        hosts.add(host)
+        hub_names.add(name)
+        return True
+
+    def validate_entity(hub_name: str, entity: dict) -> bool:
+        """Validate entity."""
+        nonlocal minimum_scan_interval, ent_names, ent_addr
+        name = entity[CONF_NAME]
+        addr = str(entity[CONF_ADDRESS])
+        scan_interval = entity.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        if scan_interval < 5:
+            _LOGGER.warning(
+                (
+                    "%s %s scan_interval(%d) is lower than 5 seconds, "
+                    "which may cause Home Assistant stability issues"
+                ),
+                hub_name,
+                name,
+                scan_interval,
+            )
+        entity[CONF_SCAN_INTERVAL] = scan_interval
+        minimum_scan_interval = min(scan_interval, minimum_scan_interval)
+        for conf_type in (
+            CONF_INPUT_TYPE,
+            CONF_WRITE_TYPE,
+            CONF_COMMAND_ON,
+            CONF_COMMAND_OFF,
+        ):
+            if conf_type in entity:
+                addr += "_" + str(entity[conf_type])
+        inx = entity.get(CONF_SLAVE, None) or entity.get(CONF_DEVICE_ADDRESS, 0)
+        addr += "_" + str(inx)
+        loc_addr: set[str] = set()
+        loc_addr.add(addr)
+
+        if CONF_TARGET_TEMP in entity:
+            a = str(entity[CONF_TARGET_TEMP])
+            a += "_" + str(inx)
+            loc_addr.add(a)
+        if CONF_HVAC_MODE_REGISTER in entity:
+            a = str(entity[CONF_HVAC_MODE_REGISTER][CONF_ADDRESS])
+            a += "_" + str(inx)
+            loc_addr.add(a)
+        if CONF_FAN_MODE_REGISTER in entity:
```

Comment:
Hi @janiversen , did you replace the CONF_FAN_MODE_REGISTER management?

Reply:
No I combined the duplicate_ methods.

Reply:
During the next period, I will combine other validators and also centralise configuration test away from the entities.

Reply:
Ok, you are optimizing.. 
I was scared having seen the  removed the following lines... :)
 `a = str(
                        entry[CONF_FAN_MODE_REGISTER][CONF_ADDRESS]
                        if isinstance(entry[CONF_FAN_MODE_REGISTER][CONF_ADDRESS], int)
                        else entry[CONF_FAN_MODE_REGISTER][CONF_ADDRESS][0]`

Reply:
Well a review means you read the whole PR, and not just pull a single line. But yes that line went out the window, due to an optimised process.

Apart from that, since the tests still work, you should have little to worry about.